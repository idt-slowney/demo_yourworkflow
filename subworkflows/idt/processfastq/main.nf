include { FASTQC as PreprocessFASTQC }  from '../../../modules/idt/fastqc/main'
include { SEQTK_SAMPLE }                from '../../../modules/idt/seqtk/sample/main'
include { CUTADAPT_EXTRACTUMI }         from '../../../modules/idt/cutadapt/extractumi/main'
include { TRIMGALORE }                  from '../../../modules/idt/trimgalore/main'
include { FASTQC as PostprocessFASTQC } from '../../../modules/idt/fastqc/main'

workflow PROCESSFASTQ {

    take:
    reads    // channel: [ val(meta), [ reads ] ]
    skip_fastqc
    skip_downsample
    umi_flag
    skip_extractumi
    read_structure
    skip_trimming

    main:
    // Instantiate empty channels
    versions = Channel.empty()
    fastqc_html = Channel.empty()
    fastqc_zip = Channel.empty()

    // Preprocessing QC -- skip_fastqc (DEFAULT: FALSE)
    if (!skip_fastqc) {
        PreprocessFASTQC(reads)
        versions = versions.mix(PreprocessFASTQC.out.versions)
        fastqc_html = fastqc_html.mix(PreprocessFASTQC.out.html)
        fastqc_zip = fastqc_zip.mix(PreprocessFASTQC.out.zip)
    }

    // Downsampling -- skip_downsample (DEFAULT: FALSE)
    // For now - we branch and mix here to account for situations with partial downsampling
    // #TODO - decide if we want to keep it like this, where downsampling is done per sample in the samplesheet
    // or set one downsampling level as a parameter for all samples - this will probably impact the conditional
    sample_reads = reads
    if (!skip_downsample) {
        downsample_reads = sample_reads.branch{
                    Sample: it[0]["sample_size"] && it[0]["sample_size"] != null
                    No_Sample: it[0]["sample_size"] == null | !("sample_size" in it[0]) }
                    
        SEQTK_SAMPLE(
            downsample_reads.Sample
            .map{ meta, reads -> [meta, reads, meta.sample_size] } )
        versions = versions.mix(SEQTK_SAMPLE.out.versions)
        sample_reads = downsample_reads.No_Sample.mix(SEQTK_SAMPLE.out.reads)
    }

    // ExtractUMI -- umi_flag (FALSE), skip_extractumi (FALSE)
    extractumi_reads = sample_reads
    if (umi_flag && !skip_extractumi) {
        CUTADAPT_EXTRACTUMI(extractumi_reads, read_structure)
        versions = versions.mix(CUTADAPT_EXTRACTUMI.out.versions)
        extractumi_reads = CUTADAPT_EXTRACTUMI.out.reads
    }

    // Trim Adapters -- skip_trimming (FALSE)
    trim_reads = extractumi_reads
    trimgalore_log          = Channel.empty()
    trimgalore_unpaired     = Channel.empty()
    trimgalore_html         = Channel.empty()
    trimgalore_zip          = Channel.empty()
    if (!skip_trimming) {
        TRIMGALORE(trim_reads)
        // Here we can include other trimming tools and control which get run with the when directive
        // i.e. TRIMMOMATIC(trim_reads) 
        // ---  { ext.when = trimmer = trimmomatic }
        // Then we mix results of all tools with a previously instantiated empty channel and yield only the output from the process that was run
        // i.e. trimmed_out = channel.empty()
        // ---  trimmed_out = trimmed_out.mix(TRIMGALORE.out.reads)
        // ---  trimmed_out = trimmed_out.mix(TRIMMOMATIC.out.reads)
        trimgalore_log          = trimgalore_log.mix(TRIMGALORE.out.log)
        trimgalore_unpaired     = trimgalore_unpaired.mix(TRIMGALORE.out.unpaired)
        trimgalore_html         = trimgalore_html.mix(TRIMGALORE.out.html)
        trimgalore_zip          = trimgalore_zip.mix(TRIMGALORE.out.zip)
        versions = versions.mix(TRIMGALORE.out.versions)
        trim_reads = TRIMGALORE.out.reads
    } 
    // Postprocessing QC
    PostprocessFASTQC(trim_reads)
    versions = versions.mix(PostprocessFASTQC.out.versions)
    fastqc_html = fastqc_html.mix(PostprocessFASTQC.out.html)
    fastqc_zip = fastqc_zip.mix(PostprocessFASTQC.out.zip)

    emit:
    reads               = trim_reads           // channel: [ val(meta), [ reads ] ]
    fastqc_html         = fastqc_html          // channel: [ val(meta), path(html) ]
    fastqc_zip          = fastqc_zip           // channel: [ val(meta), path(zip) ]
    trimgalore_log                             // channel: [ val(meta), path(log) ]
    trimgalore_unpaired                        // channel: [ val(meta), [ unpaired_reads ] ]
    trimgalore_html                            // channel: [ val(meta), path(html) ]
    trimgalore_zip                             // channel: [ val(meta), path(zip) ]
    versions = versions                     // channel: [ versions.yml ]
}

