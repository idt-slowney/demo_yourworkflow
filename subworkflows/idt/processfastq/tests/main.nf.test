// TODO nf-core: Once you have added the required tests, please run the following command to build this file:
// nf-core subworkflows test processfastq
nextflow_workflow {

    name "Test Subworkflow PROCESSFASTQ"
    script "../main.nf"
    workflow "PROCESSFASTQ"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/processfastq"
    // TODO nf-core: Add tags for all modules used within this subworkflow. Example:
    tag "fastqc"
    tag "seqtk/sample"
    tag "cutadapt/extractumi"
    tag "trimgalore"


    // TODO nf-core: Change the test name preferably indicating the test-data and file-format used
    test("homo_sapiens - fastq - paired_end") {

        when {
            workflow {
                """
                skip_fastqc        = false 
                skip_downsample    = false
                umi_flag           = false
                skip_extractumi    = true
                read_structure     = '8M143T 8M143T'
                skip_trimming      = false

                // TODO nf-core: define inputs of the workflow here. Example:
                input[0] = Channel.of([
                    [ id:'test', single_end:false, sample_size:10000 ],
                    [
                        file(params.test_data['homo_sapiens']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                        file(params.test_data['homo_sapiens']['illumina']['test_2_fastq_gz'], checkIfExists: true)
                    ]
                ])
                input[1] = skip_fastqc
                input[2] = skip_downsample
                input[3] = umi_flag
                input[4] = skip_extractumi
                input[5] = read_structure
                input[6] = skip_trimming
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.fastqc_html },
                { assert workflow.out.fastqc_zip },
                { assert snapshot(
                        workflow.out.reads,
                        workflow.out.versions
                    ).match()
                }
            )
        }
    
    // test("skip fastqc") {

    //     when {
    //         workflow {
    //             """
    //             skip_fastqc        = true 
    //             skip_downsample    = true
    //             umi_flag           = false
    //             skip_extractumi    = true
    //             read_structure     = '8M143T 8M143T'
    //             skip_trimming      = false

    //             input[0] = Channel.of([
    //                 [ id:'test', single_end:false, sample_size:10000 ], // meta map
    //                 [ file(params.test_data['homo_sapiens']['illumina']['test_1_fastq_gz'], checkIfExists: true),
    //                   file(params.test_data['homo_sapiens']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
    //             ])
    //             input[1] = skip_fastqc
    //             input[2] = skip_downsample
    //             input[3] = umi_flag
    //             input[4] = skip_extractumi
    //             input[5] = read_structure
    //             input[6] = skip_trimming
    //             """
    //         }
    //     }

    //     then {
    //         assertAll(
    //             { assert workflow.success},
    //             { assert snapshot(
    //                 workflow.out.reads,
    //                 workflow.out.versions).match() }
    //             )
    //         }
    //     }
    }
}
