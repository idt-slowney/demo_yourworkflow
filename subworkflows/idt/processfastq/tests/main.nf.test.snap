{
    "homo_sapiens - fastq - paired_end": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "sample_size": 10000
                    },
                    [
                        "test_1_val_1.fq.gz:md5,1155b07c8311d64495821fb45014e03e",
                        "test_2_val_2.fq.gz:md5,6c55135fa6eb1138d0c479e8f7dc5fae"
                    ]
                ]
            ],
            [
                "versions.yml:md5,06e5e3503bcf65395062c31d07a590f3",
                "versions.yml:md5,77e3d93e13faa7e33fcad5bc8f577f2b",
                "versions.yml:md5,7b9f999b76de4a7d95267dfd2d9ddec0",
                "versions.yml:md5,e0d4c6ee367a4d6600b7268278549bae"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-06T11:27:11.488962219"
    }
}